h2/src/main/org/h2/mvstore/tx/TransactionStore.java

```diff
@@ -245,25 +287,54 @@ public synchronized void close() {
      *
      * @return the transaction
      */
-    public synchronized Transaction begin() {
+    public Transaction begin() {
+        return begin(RollbackListener.NONE);
+    }
+
+    /**
+     * Begin a new transaction.
+     * @param listener to be notified in case of a rollback
+     *
+     * @return the transaction
+     */
+    public Transaction begin(RollbackListener listener) {
+        Transaction transaction = registerTransaction(0, Transaction.STATUS_OPEN, null, 0, listener);
+        return transaction;
+    }
+
+    private Transaction registerTransaction(int txId, int status, String name, long logId,
+                                            RollbackListener listener) {
 
         int transactionId;
-        int status;
-        if (!init) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_TRANSACTION_ILLEGAL_STATE,
-                    "Not initialized");
-        }
-        transactionId = openTransactions.nextClearBit(1);
-        if (transactionId > maxTransactionId) {
-            throw DataUtils.newIllegalStateException(
-                    DataUtils.ERROR_TOO_MANY_OPEN_TRANSACTIONS,
-                    "There are {0} open transactions",
-                    transactionId - 1);
-        }
-        openTransactions.set(transactionId);
-        status = Transaction.STATUS_OPEN;
-        return new Transaction(this, transactionId, status, null, 0);
+        long sequenceNo;
+        boolean success;
+        do {
+            VersionedBitSet original = openTransactions.get();
+            if (txId == 0) {
+                transactionId = original.nextClearBit(1);
+            } else {
+                transactionId = txId;
+                assert !original.get(transactionId);
+            }
+            if (transactionId > maxTransactionId) {
+                throw DataUtils.newIllegalStateException(
+                        DataUtils.ERROR_TOO_MANY_OPEN_TRANSACTIONS,
+                        "There are {0} open transactions",
+                        transactionId - 1);
+            }
+            VersionedBitSet clone = original.cloneIt();
+            clone.set(transactionId);
+            sequenceNo = clone.getVersion() + 1;
+            clone.setVersion(sequenceNo);
+            success = openTransactions.compareAndSet(original, clone);
+        } while(!success);
+
+        Transaction transaction = new Transaction(this, transactionId, sequenceNo, status, name, logId, listener);
+
+        success = transactions.compareAndSet(transactionId, null, transaction);
+        assert success;
```

Comment:
In this place assignment into variable looks even more confusing. Looks like this value is used in some other place, but it's not.

Reply:
Invocation of `compareAndSet()` cannot be inlined into `assert` here because `assert` excludes the whole statement from execution if assertions are not enabled, but it's not clear why result is checked with assertion.

Reply:
Of course it can’t be inlined. That was the whole point of the introduction of that variable. And it is my fear that someone might. I agree that it looks somewhat confusing, and that is because it really should be just unconditional set() call instead. Transaction slot has to be empty (null) at this point, but my level of confidence was not high enough at the time to warrant unconditional set. I should replace it after a few weeks, month, releases? :smile: